### PR TITLE
AI Chat fullpage UI notices and polish

### PR DESCRIPTION
**Pull request by petemill**
Blocked on:
- [x] https://github.com/brave/brave-core/pull/26767

## Description
- Shows a notice that the full-page feature is now available
- Shows a notice when the conversation list is empty
- Shows a notice when storage pref is disabled and offers an action to enable
- Fixes FullPage showing as SidePanel mode when restoring browser state after startup
- Optimizes loading by fetching initial properties as soon as JS is ready and not after first react render (seems to save up to 60ms on my machine, but that's in dev)
- Shows a full-page loading indicator and other minor page-load visual polish
- Fix "new conversation" button showing on collapsed navigation when conversation is already new
- AI Chat conversation items get their intended max-width and line up with input box

- Resolves https://github.com/brave/brave-browser/issues/42413
- Resolves https://github.com/brave/brave-browser/issues/42360
- Resolves https://github.com/brave/brave-browser/issues/42576

-----------------------

image

image

image

image

image

image

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by petemill** [comment]
Depends on https://github.com/brave/leo/pull/912

**Comment by diracdeltas** [comment]
does https://github.com/brave/reviews/issues/1703#issuecomment-2260662872 need to be addressed now or is this still all behind a feature flag?

File: components/ai_chat/core/browser/ai_chat_service.cc
Comment (ShivanKaul) [inline components/ai_chat/core/browser/ai_chat_service.cc:581]: We reach this if the user presses X on the top-right of the `Conversation history` popup in Leo, right? We also need an explicit "Don't do this" or "No thanks" button on that popup, otherwise this seems like a consent dark pattern. The language on that popup right now makes me think that the X will actually just dismiss the popup without actually opting me out of this feature change.

File: components/resources/ai_chat_ui_strings.grdp
Comment (ShivanKaul) [inline components/resources/ai_chat_ui_strings.grdp:274]:
```suggestion
    Leo will now remember your conversations so you can go back to them. They are stored encrypted on your device, and you can delete them any time.
```

File: components/ai_chat/resources/page/components/notices/notice_conversation_storage.tsx
Comment (ShivanKaul) [inline components/ai_chat/resources/page/components/notices/notice_conversation_storage.tsx:51]: Let's make sure the help page for this feature gets reviewed before shipping.

File: components/resources/ai_chat_ui_strings.grdp
Comment (mkarolin) [inline components/resources/ai_chat_ui_strings.grdp:274]: Maybe don't need the word `previous` here?

Comment (ShivanKaul) [reply components/resources/ai_chat_ui_strings.grdp:274]: great minds @mkarolin

File: components/resources/ai_chat_ui_strings.grdp
Comment (ShivanKaul) [inline components/resources/ai_chat_ui_strings.grdp:283]: Similarly here, I don't think we need the word "previous". Not blocking.

**Comment by ShivanKaul** [comment]
It's hard to review this from a privacy/consent perspective just from the screenshots. I suggested changes for the text, but it's not clear to me if the intention is to block users from going back to the old behaviour on sidebar Leo (i.e. ephemeral-only chats). Does the "History is disabled" prompt block users from interacting with Leo?

**Comment by petemill** [comment]
> Does the "History is disabled" prompt block users from interacting with Leo?

@ShivanKaul this prompt shows when the user has disabled the AI Chat history storage preference (which is enabled by default for all users). It does not block anything. It informs the user that conversations are ephemeral and won't show up in a list once the conversation is closed.

File: components/ai_chat/core/browser/ai_chat_service.cc
Comment (petemill) [reply components/ai_chat/core/browser/ai_chat_service.cc:581]: @ShivanKaul The storage notice is dismissed either by clicking the X or after it's been viewed once for a few seconds.
As far as I understand, the opinion was that we did not need to present the opt-out preference inline as it's not going to affect any previous actions the user does, only the future actions. And that we don't need explicit opt-in permission from the user to store submitted conversations on their own computer (encrypted) for the AI chat feature. This seems more of a notice about the new feature. @mattmcalister

File: components/resources/ai_chat_ui_strings.grdp
Comment (petemill) [reply components/resources/ai_chat_ui_strings.grdp:283]: I'd argue previous in this case is necessary (although I'll remove the "and search" part because we don't have that). Feels a bit strange to be viewing a conversation (a non-stored one) with a notice saying "In order to view your conversations with Leo, you need to...".

**Comment by petemill** [comment]
> It's hard to review this from a privacy/consent perspective just from the screenshots

@ShivanKaul In that case, I suggest we don't block this PR so that reviewers can easily use the feature end-to-end in nightly.  Although if there are obvious required changes to these pieces then we should do them here. This also doesn't contain any initial opt-in changes that may be required.

Instead we can block feature flag removal unril privacy review is successful.

File: components/resources/ai_chat_ui_strings.grdp
Comment (petemill) [reply components/resources/ai_chat_ui_strings.grdp]: done

File: components/ai_chat/resources/page/components/notices/notice_conversation_storage.tsx
Comment (petemill) [reply components/ai_chat/resources/page/components/notices/notice_conversation_storage.tsx:52]: this isn't the url it points to, it's the current Leo help page

File: components/resources/ai_chat_ui_strings.grdp
Comment (petemill) [reply components/resources/ai_chat_ui_strings.grdp]: done

File: components/ai_chat/resources/page/components/notices/notice_conversation_storage.tsx
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/components/notices/notice_conversation_storage.tsx:37]: It's not safe to use a ref in a dependency array - instead you can pass in a ref function:

```tsx
const visibilityTimer = React.useRef<VisibilityTimer | null>(null)
const noticeElementRef = React.useCallback((el: HTMLDivElement | null) => {
    // note: el will be null when we destroy it.
    // note: In new versions of React (maybe newer than we're using) you can return a cleanup function instead
    // https://react.dev/blog/2024/04/25/react-19#cleanup-functions-for-refs
    if (visibilityTimer.current) {
        visibilityTimer.current.stopTracking()
    }

    if (!el) return

    visibilityTimer.current = new VisibilityTimer(aiChatContext.markStorageNoticeViewed, 4000, el)
    visibilityTimer.current.startTracking()
}, [])

return <div ref={noticeElementRef}/>
```

File: components/ai_chat/resources/page/components/notices/notices.module.scss
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/components/notices/notices.module.scss:22]: no Nala color?

File: components/ai_chat/resources/page/components/notices/notices.module.scss
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/components/notices/notices.module.scss:26]: is the `3%` from Figma? It feels slightly odd

File: components/ai_chat/resources/page/components/notices/notices.module.scss
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/components/notices/notices.module.scss:63]: optional: Use Nala spacing tokens in this file

File: components/ai_chat/resources/page/state/ai_chat_context.tsx
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/state/ai_chat_context.tsx:18]: Its a bit weird that this is `mojom.UIState` when the `UIState` interface isn't really part of the mojom API but a thing we're using to wrap some state

File: components/ai_chat/resources/page/state/ai_chat_context.tsx
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/state/ai_chat_context.tsx:70]: nit: capital `U` `UIState` reads slightly weird to me - normally that'd be the name of a class but here its a bag of data

File: components/ai_chat/resources/page/state/ai_chat_context.tsx
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/state/ai_chat_context.tsx:71]: looking at this, does using the `CustomEvent` API buy as anything except needing to access the data through `.detail`? We don't need it to bubble or anything (and it doesn't)

File: components/ai_chat/resources/page/state/ai_chat_context.tsx
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/state/ai_chat_context.tsx:83]: optional
```suggestion
    markStorageNoticeViewed: Service.dismissStorageNotice,
```

File: components/ai_chat/resources/page/state/ai_chat_context.tsx
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/state/ai_chat_context.tsx:85]: just checking, but this is just eagerly changing the data right? After the "backend" does the update we'll receive a notification anyway

File: components/ai_chat/resources/page/ai_chat_ui.html
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/ai_chat_ui.html:19]: any reason we don't just load the Nala CSS variables and use it directly here?

File: components/ai_chat/resources/page/ai_chat_ui.html
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/ai_chat_ui.html:33]: could we keep the type here as `module` and add the flag to the BUILD.gn to output a module? I think we should be gradually migrating anyway

File: components/resources/ai_chat_ui_strings.grdp
Comment (fallaciousreasoning) [inline components/resources/ai_chat_ui_strings.grdp:271]: This string already exists here (for the title of the conversation history list). Any chance we could share them?

File: package.json
Comment (fallaciousreasoning) [inline package.json:181]: Any chance we can bump Nala separately, so we can generate a changelog PR?

File: package-lock.json
Comment (fallaciousreasoning) [inline package-lock.json:152]: is this intentional?

File: components/ai_chat/resources/page/chat_ui.tsx
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/chat_ui.tsx:26]: does this mean we have to be a bit more careful about calling `setAPIForTesting` early to not break Storybook?

**Review by fallaciousreasoning — COMMENTED** [review]
This is looking pretty good. Stepping back a little bit, I wonder if we should have a generic `StateProvider` context that we can just provide some data to and do all the logic in the API wrapper?

i.e. we have some kind of interface our API wrappers conform to and then we have something to do the Context boilerplate automatically?

That'd solve main main problem with Kevin's current solution, and keep it so we can override the state in Storybook. WDYT?

File: components/ai_chat/resources/page/components/conversations_list/index.tsx
Comment (petemill) [reply components/ai_chat/resources/page/components/conversations_list/index.tsx:138]: I _normally_ don't indent if it's not DOM structure related

File: components/ai_chat/resources/page/components/notices/notices.module.scss
Comment (petemill) [reply components/ai_chat/resources/page/components/notices/notices.module.scss:22]: I don't think so, and it's not meant to change in light/dark anyway

File: components/ai_chat/resources/page/components/notices/notices.module.scss
Comment (petemill) [reply components/ai_chat/resources/page/components/notices/notices.module.scss:26]: The calculation at 680px is, yes. Felt right to me to vary the padding based on width

**Review by mkarolin — APPROVED** [review]
`strings`++

File: components/ai_chat/resources/page/components/conversations_list/index.tsx
Comment (fallaciousreasoning) [reply components/ai_chat/resources/page/components/conversations_list/index.tsx:138]: Fair enough - in my head I normally just think of the JSX as Javascript statements and if I had 

```tsx
foo &&
  bar
```

I'd indent the second bit, so I'll do the same with the JSX

Let's leave it and make prettier work properly (for realsies) :smile:

File: components/ai_chat/resources/page/chat_ui.tsx
Comment (petemill) [reply components/ai_chat/resources/page/chat_ui.tsx:26]: `setAPIForTesting` isn't used (we should delete the function!) - storybook simply provides its own `AIChatContext`. And since AIChatContext extends `UIState`, storybook provides it's own version for that

File: package-lock.json
Comment (petemill) [reply package-lock.json:152]: no, it happened as a result of the temporary leo update

File: package.json
Comment (petemill) [reply package.json:181]: yes, this is temporary (see commits)

File: components/ai_chat/resources/page/ai_chat_ui.html
Comment (petemill) [reply components/ai_chat/resources/page/ai_chat_ui.html:33]: Should the main bundle be module? The reason I removed it was to have it load as eagerly as possible, but maybe I'm mistaken?

File: components/ai_chat/resources/page/ai_chat_ui.html
Comment (petemill) [reply components/ai_chat/resources/page/ai_chat_ui.html:19]: To avoid a FOUC ahead of the nala css loading

File: components/ai_chat/resources/page/components/notices/notice_conversation_storage.tsx
Comment (petemill) [reply components/ai_chat/resources/page/components/notices/notice_conversation_storage.tsx:13]: My muscle memory is a bit rusty!

File: components/ai_chat/resources/page/api/index.ts
Comment (petemill) [reply components/ai_chat/resources/page/api/index.ts:113]: My concern is that `API.addListener` seems like it's listening for any/all of the mojom API observer functions instead of a specific piece of state

File: components/ai_chat/core/browser/ai_chat_service.cc
Comment (petemill) [reply components/ai_chat/core/browser/ai_chat_service.cc:568]: Agreed - over the course of this PR it grew

File: components/ai_chat/resources/page/api/index.ts
Comment (petemill) [reply components/ai_chat/resources/page/api/index.ts]: and it probably did when part of AIChatContext

File: components/ai_chat/resources/page/api/index.ts
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/api/index.ts:47]: :+1:

**Review by fallaciousreasoning — APPROVED** [review]
lgtm

**Review by ShivanKaul — APPROVED** [review]
We can do an overall privacy review after merge and before the feature flag is turned on by default

File: components/ai_chat/core/browser/ai_chat_service_unittest.cc
Comment (petemill) [reply components/ai_chat/core/browser/ai_chat_service_unittest.cc:77]: the callback via mojom interface is not optional

**Comment by petemill** [comment]
> @petemill [Re](https://github.com/brave/brave-core/pull/26678#issuecomment-2494631379) How do users disable history? How are we notifying them?

@PrivacyMatters users can disable storage of conversations and being able to continue them in browser Settings. It will be explained on the "Learn more" link that will show in the notice announcing storage for all users - existing and new.

File: components/ai_chat/resources/page/ai_chat_ui.html
Comment (fallaciousreasoning) [reply components/ai_chat/resources/page/ai_chat_ui.html:33]: I think in this scenario, yes. So `<script type=module>` is `defer` by default so it will execute after the HTML has finished parsing (but before `DOMContentLoaded` fires), but it will start loading as soon as the parser sees the tag in the document.

I think we should put the `<script type=module src=..>` in the head of the document. Using the raw script tag will block further parsing of the HTML file until parsing of the script completes (though this is at the end of the document so its probably fine).

Either way, I don't think there should be a huge difference.

https://jakearchibald.com/2017/es-modules-in-browsers/

File: components/ai_chat/resources/page/ai_chat_ui.html
Comment (fallaciousreasoning) [reply components/ai_chat/resources/page/ai_chat_ui.html:19]: in the head of the document, right?

File: package-lock.json
Comment (petemill) [reply package-lock.json:152]: (and because someone has modified `engines` in `package.json` but not `package-lock.json`)

File: components/ai_chat/resources/page/components/code_block/index.tsx
Comment (petemill) [reply components/ai_chat/resources/page/components/code_block/index.tsx:13]: ah, I changed one of the imports because there's no reason to be importing from `brave-ui` so I cleaned up the very-wrong order. But I moved them a little too far. Fixing

File: components/ai_chat/resources/page/components/conversations_list/style.module.scss
Comment (petemill) [reply components/ai_chat/resources/page/components/conversations_list/style.module.scss:7]: Actually no, the issue is that the parent width animates closed causing the conversation titles to weirdly collapse. I think it's a better animation if it stays full-width whilst collapsing. Ideally this wix would be fixed on the relevant parent and contained within FullPage's styles. On reflection, I did that!

File: components/ai_chat/resources/page/components/loading/index.tsx
Comment (petemill) [reply components/ai_chat/resources/page/components/loading/index.tsx:10]: Let's think about it - at least we're sharing ProgressRing

File: components/ai_chat/resources/page/components/main/index.tsx
Comment (petemill) [reply components/ai_chat/resources/page/components/main/index.tsx:211]: For your trouble I've sneakily made most of the formatting consistent (in this file)

File: components/ai_chat/resources/page/components/main/style.module.scss
Comment (petemill) [reply components/ai_chat/resources/page/components/main/style.module.scss:7]: I've converted this and a few others in this file to leo spacing variables

File: components/ai_chat/resources/page/components/notices/notices.module.scss
Comment (petemill) [reply components/ai_chat/resources/page/components/notices/notices.module.scss]: done

File: components/ai_chat/resources/page/ai_chat_ui.html
Comment (petemill) [reply components/ai_chat/resources/page/ai_chat_ui.html:19]: lol what is happening with my comments these days!

File: components/resources/ai_chat_ui_strings.grdp
Comment (petemill) [reply components/resources/ai_chat_ui_strings.grdp:277]: I changed it. Either way I think it only gets translated once, but the other key makes some semantic sense here too so why not

**Review by Brandon-T — APPROVED** [review]

File: ios/brave-ios/Sources/AIChat/ModelView/AIChatViewModel.swift
Comment (petemill) [inline ios/brave-ios/Sources/AIChat/ModelView/AIChatViewModel.swift:75]: @Brandon-T we don't need to call `objectWillChange.send()` right?

**Review by Brandon-T — APPROVED** [review]

File: ios/brave-ios/Sources/AIChat/ModelView/AIChatViewModel.swift
Comment (Brandon-T) [reply ios/brave-ios/Sources/AIChat/ModelView/AIChatViewModel.swift:75]: Correct. Not necessary.

**Pull request merged**
